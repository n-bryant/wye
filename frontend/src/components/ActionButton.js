import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

import styles from "./ActionButton.styles";

/**
 * renders an animated Button with a given label and click handler
 */
export const ActionButton = props => {
  const classnames = ActionButton.classnames(props);
  const { label, onClick, disabled } = props;
  return (
    <div className={classnames.root()}>
      <Button
        className={classnames.element("button")}
        onClick={onClick}
        disabled={disabled}
      >
        <svg
          className={classnames.element("buttonBorder", { disabled })}
          width="180px"
          height="60px"
          viewBox="0 0 180 60"
        >
          <polyline points="179,1 179,59 1,59 1,1 179,1" />
          <polyline points="179,1 179,59 1,59 1,1 179,1" />
        </svg>
        <Typography variant="body1">{label}</Typography>
      </Button>
    </div>
  );
};
ActionButton.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}ActionButton`
);
ActionButton.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    button: PropTypes.string,
    buttonBorder: PropTypes.string,
    buttonBorderDisabled: PropTypes.string
  }),
  // the label text for the button
  label: PropTypes.string,
  // the click handler for the button
  onClick: PropTypes.func,
  // whether the button should be disabled
  disabled: PropTypes.bool
};
ActionButton.defaultProps = {
  classes: {},
  label: "Go!",
  onClick: () => {}
};

export default withStyles(styles)(ActionButton);
