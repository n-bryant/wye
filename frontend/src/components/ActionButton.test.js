import { ActionButton } from "./ActionButton";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

describe("ActionButton", () => {
  const props = {
    classes: {
      root: "root",
      button: "button",
      buttonBorder: "buttonBorder",
      buttonBorderDisabled: "buttonBorderDisabled"
    },
    label: "foo",
    onClick: jest.fn(),
    disabled: false
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ActionButton {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should call the onClick prop when clicked", () => {
    wrapper.find(Button).simulate("click");
    expect(props.onClick).toHaveBeenCalled();
  });

  it("should set disabled styles if the disabled prop is truthy", () => {
    // falsy disabled value
    expect(
      wrapper.findWhere(n => n.hasClass(props.classes.buttonBorderDisabled))
        .length
    ).toBe(0);

    // truthy disabled value
    wrapper.setProps({
      disabled: true
    });
    wrapper.update();
    expect(
      wrapper.findWhere(n => n.hasClass(props.classes.buttonBorderDisabled))
        .length
    ).toBe(1);
  });

  it("should set the button's text to the value of the label prop", () => {
    expect(wrapper.find(Typography).text()).toBe(props.label);
  });
});
