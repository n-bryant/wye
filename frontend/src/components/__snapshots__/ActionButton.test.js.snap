// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton should render successfully 1`] = `
<div
  className="wye-ActionButton root"
>
  <WithStyles(ForwardRef(Button))
    className="wye-ActionButton__button button"
    disabled={false}
    onClick={[MockFunction]}
  >
    <svg
      className="wye-ActionButton__buttonBorder buttonBorder"
      height="60px"
      viewBox="0 0 180 60"
      width="180px"
    >
      <polyline
        points="179,1 179,59 1,59 1,1 179,1"
      />
      <polyline
        points="179,1 179,59 1,59 1,1 179,1"
      />
    </svg>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      foo
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Button))>
</div>
`;
